File: src/main/ts/booklet/impure/router/BookRouter.ts
import { Http, Request, Response } from 'some-http-library';
import { BookHandler } from '../service/BookHandler';
import { Failure } from '../../pure/Failure';
import { CustomResponse } from '../../pure/http/CustomResponse';

export const BookRouter = {
  app: Http.collectZIO<Request, Response>({
    'GET /books': async (req: Request) => {
      const userId = req.headers['user_id'];
      if (!userId) {
        return CustomResponse.badRequest('Missing user ID');
      }
      try {
        const response = await BookHandler.fetchAll(userId);
        return response;
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    },
    'GET /books/:bookId': async (req: Request) => {
      const userId = req.headers['user_id'];
      const bookId = req.params.bookId;
      if (!userId || !bookId) {
        return CustomResponse.badRequest('Missing user ID or book ID');
      }
      try {
        const response = await BookHandler.fetch(bookId, userId);
        return response;
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    },
    'POST /books': async (req: Request) => {
      const userId = req.headers['user_id'];
      if (!userId) {
        return CustomResponse.badRequest('Missing user ID');
      }
      try {
        const response = await BookHandler.create(req, userId);
        return response;
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    },
    'PATCH /books/:bookId': async (req: Request) => {
      const userId = req.headers['user_id'];
      const bookId = req.params.bookId;
      if (!userId || !bookId) {
        return CustomResponse.badRequest('Missing user ID or book ID');
      }
      try {
        const response = await BookHandler.update(bookId, req, userId);
        return response;
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    },
    'DELETE /books/:bookId': async (req: Request) => {
      const userId = req.headers['user_id'];
      const bookId = req.params.bookId;
      if (!userId || !bookId) {
        return CustomResponse.badRequest('Missing user ID or book ID');
      }
      try {
        const response = await BookHandler.delete(bookId, userId);
        return response;
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    }
  })
};

File: src/main/scala/booklet/impure/router/BookRouter.scala
package booklet.impure.router

import booklet.impure.service.BookHandler
import booklet.pure.Failure
import booklet.pure.http.CustomResponse.serverFailure
import zhttp.http.Method.{DELETE, GET, PATCH, POST}
import zhttp.http._

object BookRouter {
  val app: Http[BookHandler, Nothing, Request, Response] =
    Http
      .collectZIO[Request] {
        case req @ GET -> !! / "books"             => BookHandler.fetchAll(req)
        case req @ GET -> !! / "books" / bookId    => BookHandler.fetch(req, bookId)
        case req @ POST -> !! / "books"            => BookHandler.create(req)
        case req @ PATCH -> !! / "books" / bookId  => BookHandler.update(req, bookId)
        case req @ DELETE -> !! / "books" / bookId => BookHandler.delete(req, bookId)
      }
      .foldHttp(
        failure => Http.succeed(serverFailure(failure)),
        defect => Http.succeed(serverFailure(Failure.fromThrowable(defect))),
        success => Http.succeed(success),
        Http.empty
      )
}
